### PR TITLE
[20325][20701] Replace dynamic finders, fix ActiveRecord deprecation warnings

### PR DESCRIPTION
**Pull request by myabc**
:information_desk_person: **This PR doesn't deal with deprecation warnings that are not ActiveRecord related.** 

Supercedes #3186

https://community.openproject.org/work_packages/20325
https://community.openproject.org/work_packages/20701

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://myabc-feature-20325-rails4-deprecated-dynamic-finders.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://myabc-feature-20325-rails4-deprecated-dynamic-finders.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

File: lib/plugins/acts_as_tree/test/acts_as_tree_test.rb [hunk 1]
```diff
@@ -181,17 +181,17 @@ def test_eager_association_loading
...
 
   def test_eager_association_loading_with_recursive_cascading_three_levels_has_many
-    root_node = RecursivelyCascadedTreeMixin.find(:first, include: { children: { children: :children } }, order: 'mixins.id')
+    root_node = RecursivelyCascadedTreeMixin.includes({ children: { children: :children } }).order('mixins.id').first
     assert_equal @rc4, assert_no_queries { root_node.children.first.children.first.children.first }
   end
 
   def test_eager_association_loading_with_recursive_cascading_three_levels_has_one
-    root_node = RecursivelyCascadedTreeMixin.find(:first, include: { first_child: { first_child: :first_child } }, order: 'mixins.id')
+    root_node = RecursivelyCascadedTreeMixin.includes({ first_child: { first_child: :first_child } }).order('mixins.id').first
```
Comment (myabc) [inline lib/plugins/acts_as_tree/test/acts_as_tree_test.rb:189]: @ulferts apparently these plugin tests are never executed. shouldn't we move them?

**Comment by myabc** [comment]
`Api::V2::UsersController` seems to be flickering:

```
be rspec ./spec/controllers/api/v2/users_controller_spec.rb:45 --seed 61336
be rspec ./spec/controllers/api/v2/users_controller_spec.rb:45 --seed 56166
```

**Comment by myabc** [comment]
@opf/developers please begin reviewing. I realise the code reformatting complicates review somewhat (but necessary, so that I can keep re-running rubocop on changes).

57618b2 and [e02eb01...feature/20325-rails4-deprecated-dynamic-finders](https://github.com/myabc/openproject/compare/e02eb01...feature/20325-rails4-deprecated-dynamic-finders) are the commit ranges of interest.

**Comment by myabc** [comment]
https://community.openproject.org/work_packages/20701

**Comment by myabc** [comment]
`acts_as_searchable` still needs to be upgraded, but that probably needs _extra-careful review_. I may do it as a separate PR, once this is merged.

**Comment by myabc** [comment]
@oliverguenther fyi, the code reformatting in this PR will probably lead to some conflicts with #3196.

**Comment by myabc** [comment]
@NobodysNightmare we could define our own `allow_non_nil` or `allow_unless_nil` method, but I'm not sure how common a use-case this really is.

**Comment by myabc** [comment]
Please review me :smile:

**Comment by NobodysNightmare** [comment]
Regarding the commits "Fix syntax (w/Rubocop) in XXX":

I had a rough look over the first two of those commits and initially commented inline, but to make that more general:
- Sometimes the recognition of "return value based" vs "side effect based" blocks fails
- Some "empty" (or default-ish) else branches are being removed, though they seem to serve the purpose of readability
- **tl; dr:** sometimes an automatic tool just fails to do the right thing

I am not sure how we want to deal with this problem in general. Arguably these changes by you make the overall situation better, while introducing collateral damage. That is why I am inclined to merge those changes regardless.
But I don't feel like we will be able to keep "100% `rubocop -a` compatibility" in the long term. I don't know if this is a problem...

**Comment by NobodysNightmare** [comment]
I did the same for your AR finder commits (looking at the first two and extrapolating).

The only thing that leaves me uncertain is if it is relevant in which order active record scopes are applied. My google capabilities were not good enough [to find that out](https://www.google.de/search?q=ar+scope+order&source=lnms&tbm=isch&sa=X&ei=Ju2bVbv8DMTgywOCsYDgCg&ved=0CAgQ_AUoAg&biw=1246&bih=944) ^^

I will retry to find out more about this question, but you will probably already have an answer.

**Comment by NobodysNightmare** [comment]
You should see my hint on calling `Project.allowed_to_condition`, `includes` and `references` as a more general advice. I've seen this pattern with other methods too, where the caller uses some kind of condition he really doesn't know, but for some reason starts to `include` and `reference` fancy tables. Sometimes acompanied with a comment that excuses for this.

I think since the introduction of `references` only amplifies that (already existing) problem, we should at least start to look for a way out for the worst offenders.

**Comment by NobodysNightmare** [comment]
Pew... reviewed until the latest commit:

> Revert "DRY WorkPackageSchema#available_custom_fields spec" … `0436446`

Looking nice (and tedious) :)

**Comment by myabc** [comment]
@NobodysNightmare 

> Pew... reviewed until the latest commit:

Thanks!

> Looking nice (and tedious) :)

Yes, sure is tedious.

I think I've now dealt with all of your comments, with the exception of reordering `references` in query chains. As we already discussed, there may be some `references` that can be removed once we're on Rails 4.1.

> tl; dr: sometimes an automatic tool just fails to do the right thing
> I am not sure how we want to deal with this problem in general. Arguably these changes by you make the overall situation better, while introducing collateral damage. That is why I am inclined to merge those changes regardless.
> But I don't feel like we will be able to keep "100% rubocop -a compatibility" in the long term. I don't know if this is a problem...

I reported one of the issues upstream (https://github.com/bbatsov/rubocop/issues/2021). I tend to be optimistic though, and think `rubocop` can only get better at linting/auto-correcting code.  It's annoying though that Hound is still on a slightly older version of `rubocop` and that there are some discrepancies in configurations.

File: app/controllers/api/v2/planning_elements_controller.rb [hunk 2]
```diff
@@ -164,7 +165,7 @@ def find_all_projects_by_project_id
           # WTF. Why do we completely skip rewiring in this case and always provide parent_ids?
           # This is totally inconistent.
           identifiers = params[:ids].split(/,/).map(&:strip)
-          @planning_elements = WorkPackage.visible(User.current).find_all_by_id(identifiers)
+          @planning_elements = WorkPackage.visible(User.current).where(id: identifiers)
```
Comment (ulferts) [inline app/controllers/api/v2/planning_elements_controller.rb:168]: This now returns an ActiveRecord::Relation. Don't know if that is a problem, though.

File: app/models/version.rb [hunk 3]
```diff
@@ -98,7 +99,10 @@ def estimated_hours
 
   # Returns the total reported time for this version
   def spent_hours
-    @spent_hours ||= TimeEntry.sum(:hours, include: :work_package, conditions: ["#{WorkPackage.table_name}.fixed_version_id = ?", id]).to_f
+    @spent_hours ||= TimeEntry.includes(:work_package)
+                     .where(["#{WorkPackage.table_name}.fixed_version_id = ?", id])
```
Comment (ulferts) [inline app/models/version.rb:103]: This could be better written as:
`where(work_package: { fixed_version_id: 5 })`

Don't know if it is in the scope of this PR, though.

File: app/models/wiki.rb [hunk 4]
```diff
@@ -110,7 +115,9 @@ def self.titleize(title)
   end
 
   def create_menu_item_for_start_page
-    wiki_menu_item = wiki_menu_items.find_or_initialize_by_title start_page, name: 'Wiki'
+    wiki_menu_item = wiki_menu_items.find_or_initialize_by(title: start_page) { |item|
+      item.name = 'Wiki'
```
Comment (ulferts) [inline app/models/wiki.rb:119]: I am aware that this is not introduced in this PR but why don't we use I18n here?

File: app/controllers/copy_projects_controller.rb [hunk 5]
```diff
@@ -75,10 +75,9 @@ def copy_project
   private
 
   def prepare_for_copy_project
-    @issue_custom_fields = WorkPackageCustomField.find(:all, order: "#{CustomField.table_name}.position")
+    @issue_custom_fields = WorkPackageCustomField.order("#{CustomField.table_name}.position")
     @types = ::Type.all
-    @root_projects = Project.find(:all,
-                                  conditions: "parent_id IS NULL AND status = #{Project::STATUS_ACTIVE}",
-                                  order: 'name')
+    @root_projects = Project.where("parent_id IS NULL AND status = #{Project::STATUS_ACTIVE}")
```
Comment (ulferts) [inline app/controllers/copy_projects_controller.rb:80]: <pre>
where(parent_id: nil, status: Project::STATUS_ACTIVE)
</pre>


would be better still.

File: app/models/version.rb (hunk 3 above)
Comment (myabc) [inline app/models/version.rb:103]: @ulferts even though it's desirable, I don't consider it in scope for this PR (or even the Rails 4 Upgrade project, as a whole).

File: app/models/wiki.rb (hunk 4 above)
Comment (myabc) [inline app/models/wiki.rb:119]: @ulferts I can't answer that :(

File: app/controllers/copy_projects_controller.rb (hunk 5 above)
Comment (myabc) [inline app/controllers/copy_projects_controller.rb:80]: @ulferts sure, I agree, but … https://github.com/opf/openproject/pull/3194#discussion_r34335331

File: app/controllers/api/v2/planning_elements_controller.rb (hunk 2 above)
Comment (myabc) [inline app/controllers/api/v2/planning_elements_controller.rb:168]: @ulferts that shouldn't be a problem.

File: app/controllers/api/v2/projects_controller.rb [hunk 6]
```diff
@@ -89,8 +87,9 @@ def planning_element_custom_fields
       protected
 
       def find_project
-        @project = Project.find params[:id],
-                                include: [{ custom_values: [{ custom_field: :translations }] }]
+        # FIXME: Rails 4.1 Upgrade: restore includes
+        #        include: [{ custom_values: [{ custom_field: :translations }] }]
+        @project = Project.find params[:id]
```
Comment (ulferts) [inline app/controllers/api/v2/projects_controller.rb:92]: Writing this again to ensure that https://github.com/myabc/openproject/commit/51512f15223b3149b409c242c677907dc2047650#commitcomment-12101597 is seen.

**Comment by ulferts** [comment]
@myabc I am done with my review and once I received statements to my comments, willing to merge this PR.

Thanks for the huge amount of work you have put in here!

File: app/controllers/api/v2/projects_controller.rb (hunk 6 above)
Comment (myabc) [inline app/controllers/api/v2/projects_controller.rb:92]: @ulferts seen and commented :) https://github.com/myabc/openproject/commit/51512f15223b3149b409c242c677907dc2047650#commitcomment-12104806

Comment (myabc) [inline app/controllers/api/v2/projects_controller.rb:92]: #3235
